### PR TITLE
Add support for Razer Blade 16 (2024)

### LINKED ISSUE #2243: Support for Razer Blade 16 (2024)
### Device Info

* Name: Razer Blade 16 2024
* Product Number: RZ09-05102EJ9
* Model Number: RZ09-0510

### Device Information
```
Bus 001 Device 004: ID 1532:02b7 Razer USA, Ltd Razer Blade
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 [unknown]
  bDeviceSubClass         0 [unknown]
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x02b7 Razer Blade
  bcdDevice            2.00
  iManufacturer           1 Razer
  iProduct                2 Razer Blade
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x006d
    bNumInterfaces          4
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 [unknown]
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     159
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0010  1x 16 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 [unknown]
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 [unknown]
      bInterfaceProtocol      1 Keyboard
      iInterface              2 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     327
          Report Descriptors: 
            ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0002  1x 2 bytes
        bInterval               1
```

```
lrwxrwxrwx 1 root root   9 Apr 29 09:32 usb-Razer_Razer_Blade-event-if01 -> ../event8
lrwxrwxrwx 1 root root   9 Apr 29 09:32 usb-Razer_Razer_Blade-event-kbd -> ../event6
lrwxrwxrwx 1 root root   9 Apr 29 09:32 usb-Razer_Razer_Blade-if01-event-kbd -> ../event7
lrwxrwxrwx 1 root root   9 Apr 29 09:32 usb-Razer_Razer_Blade-if02-event-mouse -> ../event9
lrwxrwxrwx 1 root root   9 Apr 29 09:32 usb-Razer_Razer_Blade-if02-mouse -> ../mouse2
```

<details>

```
001:004:003:DESCRIPTOR         1714349383.341634
 05 59 09 01 A1 01 85 01 09 02 A1 02 09 03 15 00
 27 FF FF 00 00 75 10 95 01 B1 03 09 04 09 05 09
 06 09 07 09 08 15 00 27 FF FF FF 7F 75 20 95 05
 B1 03 C0 85 02 09 20 A1 02 09 21 15 00 27 FF FF
 00 00 75 10 95 01 B1 02 C0 85 03 09 22 A1 02 09
 21 15 00 27 FF FF 00 00 75 10 95 01 B1 02 09 23
 09 24 09 25 09 27 09 26 15 00 27 FF FF FF 7F 75
 20 95 05 B1 02 09 28 09 29 09 2A 09 2B 09 2C 09
 2D 15 00 26 FF 00 75 08 95 06 B1 02 C0 85 04 09
 50 A1 02 09 03 09 55 15 00 25 08 75 08 95 02 B1
 02 09 21 15 00 27 FF FF 00 00 75 10 95 08 B1 02
 09 51 09 52 09 53 09 54 09 51 09 52 09 53 09 54
 09 51 09 52 09 53 09 54 09 51 09 52 09 53 09 54
 09 51 09 52 09 53 09 54 09 51 09 52 09 53 09 54
 09 51 09 52 09 53 09 54 09 51 09 52 09 53 09 54
 15 00 26 FF 00 75 08 95 20 B1 02 C0 85 05 09 60
 A1 02 09 55 15 00 25 08 75 08 95 01 B1 02 09 61
 09 62 15 00 27 FF FF 00 00 75 10 95 02 B1 02 09
 51 09 52 09 53 09 54 15 00 26 FF 00 75 08 95 04
 B1 02 C0 85 06 09 70 A1 02 09 71 15 00 25 01 75
 08 95 01 B1 02 C0 C0

001:004:002:DESCRIPTOR         1714349383.342059
 05 01 09 02 A1 01 09 01 A1 00 05 09 19 01 29 05
 15 00 25 01 75 01 95 05 81 02 75 01 95 03 81 03
 06 00 FF 09 40 75 08 95 02 15 81 25 7F 81 02 05
 01 09 38 15 81 25 7F 75 08 95 01 81 06 09 30 09
 31 16 00 80 26 FF 7F 75 10 95 02 81 06 C0 06 00
 FF 09 02 15 00 25 01 75 08 95 5A B1 01 C0

001:004:001:DESCRIPTOR         1714349383.342731
 05 01 09 06 A1 01 85 01 05 07 19 E0 29 E7 15 00
 25 01 75 01 95 08 81 02 19 00 2A FF 00 15 00 26
 FF 00 75 08 95 0E 81 00 C0 05 0C 09 01 A1 01 85
 02 19 00 2A 3C 02 15 00 26 3C 02 95 01 75 10 81
 00 75 08 95 0D 81 01 C0 05 01 09 80 A1 01 85 03
 19 81 29 83 15 00 25 01 75 01 95 03 81 02 95 05
 81 01 75 08 95 0E 81 01 C0 05 01 09 00 A1 01 85
 04 09 03 15 00 26 FF 00 35 00 46 FF 00 75 08 95
 0F 81 00 C0 05 01 09 00 A1 01 85 05 09 03 15 00
 26 FF 00 35 00 46 FF 00 75 08 95 0F 81 00 C0

001:004:000:DESCRIPTOR         1714349383.342986
 05 01 09 06 A1 01 05 07 19 E0 29 E7 15 00 25 01
 75 01 95 08 81 02 81 01 19 00 2A FF 00 15 00 26
 FF 00 75 08 95 06 81 00 05 08 19 01 29 03 15 00
 25 01 75 01 95 03 91 02 95 05 91 01 C0
```

</details>

### PR DESCRIPTION
**Pull request by dberlin**
This PR adds support for the Razer Blade 16 (2024).
The device ID is 1532:02B7

Fixes #2243

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -106,6 +106,7 @@ The devices below are fully feature supported by OpenRazer, which means all avai
 | Razer Blade 18 (2023)                         |  1532:02A0  |
 | Razer Ornata V3 X                             |  1532:02A2  |
 | Razer BlackWidow V3 Tenkeyless                |  1532:0A24  |
+| Razer Blade 16 (2024)                         |  1532:02B7  |
```
Comment (z3ntu) [inline README.md:109]: Please move this one up so it's sorted by USB ID

File: driver/razerkbd_driver.h [hunk 2]
```diff
@@ -103,6 +103,7 @@
 #define USB_DEVICE_ID_RAZER_BLADE_18_2023 0x02A0
 #define USB_DEVICE_ID_RAZER_ORNATA_V3_X_ALT 0x02A2
 #define USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_TK 0x0A24
+#define USB_DEVICE_ID_RAZER_BLADE_16_2024 0x02B7
```
Comment (z3ntu) [inline driver/razerkbd_driver.h:106]: move this one up so it's sorted by value

File: install_files/udev/99-razer.rules [hunk 3]
```diff
@@ -10,7 +10,7 @@ ATTRS{idProduct}=="0013|0016|0020|0024|0025|0029|002e|002f|0032|0034|0036|0037|0
     ENV{ID_RAZER_CHROMA}="1", ENV{RAZER_DRIVER}="razermouse"
 
 # Keyboards
-ATTRS{idProduct}=="010d|010e|010f|0111|0113|0118|011a|011b|011c|0201|0202|0203|0204|0205|0207|0208|0209|020f|0210|0211|0214|0216|0217|021a|021e|021f|0220|0221|0224|0225|0226|0227|0228|022a|022b|022c|022d|022f|0232|0233|0234|0235|0237|0239|023a|023b|023f|0240|0241|0243|0245|0246|024a|024c|024d|024e|0252|0253|0255|0256|0257|0258|0259|025a|025d|025e|0266|0269|026a|026b|026c|026d|026e|026f|0270|0271|0276|0279|027a|0282|028a|028b|028c|028d|0290|0292|0294|0295|0296|0298|029e|029f|02a0|02a2|0a24", \
+ATTRS{idProduct}=="010d|010e|010f|0111|0113|0118|011a|011b|011c|0201|0202|0203|0204|0205|0207|0208|0209|020f|0210|0211|0214|0216|0217|021a|021e|021f|0220|0221|0224|0225|0226|0227|0228|022a|022b|022c|022d|022f|0232|0233|0234|0235|0237|0239|023a|023b|023f|0240|0241|0243|0245|0246|024a|024c|024d|024e|0252|0253|0255|0256|0257|0258|0259|025a|025d|025e|0266|0269|026a|026b|026c|026d|026e|026f|0270|0271|0276|0279|027a|0282|028a|028b|028c|028d|0290|0292|0294|0295|0296|0298|029e|029f|02a0|02a2|0a24|02b7", \
```
Comment (z3ntu) [inline install_files/udev/99-razer.rules:13]: also keep this sorted, move before 0a24

**Comment by z3ntu** [comment]
Would also be nice if you could still provide some info from https://github.com/openrazer/openrazer/blob/master/.github/ISSUE_TEMPLATE/new-device-support.md?plain=1 (ignore screenshots and packet captures).

Apart from the minor comments, looks good!

**Comment by dberlin** [comment]
Thanks. I'll fix these up later this weekend and push an updated PR

**Comment by flexa-rob** [comment]
@dberlin  - I've created a new issue with my device info. Let me know if you need any assistance/testing?

**Comment by dberlin** [comment]
Hey, it turns out I won't get to this for a while, so if you want to
take it over, feel free.

**Comment by z3ntu** [comment]
@flexa-rob If you can test this PR on your device and see if everything works that'd also be good: https://github.com/openrazer/openrazer/wiki/Building-a-package

The code style issues I commented on I can also fix when merging.

**Comment by flexa-rob** [comment]
Sure, I'll try test this branch out on my laptop (hopefully I'll find some time over the next few days).

**Comment by flexa-rob** [comment]
@z3ntu - I'm just trying to open a fresh branch, but seems I don't have access. First tried cloning @dberlin 's fork, but didn't have access either, so just copied all his changes to my local branch with all the latest changes.

```
$ git push --set-upstream origin blade-16-2024
ERROR: Permission to openrazer/openrazer.git denied to flexa-rob.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

I was also able to build/install the `.deb` packages, but was unable to list my device (does this require a daemon restart, or reboot?)

**Comment by z3ntu** [comment]
@flexa-rob You can't push to someone else's fork, if you want to push the changes somewhere you need to make a fork from your account and push it there (and open a PR from your fork then).

But do you have any changes to this PR in the first place? Not sure what you're trying to push.

> I was also able to build/install the .deb packages, but was unable to list my device (does this require a daemon restart, or reboot?)

You need to reboot the computer before it can work.

**Comment by flexa-rob** [comment]
ok, no worries @z3ntu - I'll create a fork and open a PR from there.

After rebooting, the device is displaying correctly in Polychromatic:
image

I have tested the following keyboard functions using Polychromatic:

- Brightness: works
- Off: works
- Spectrum: works
- Wave: works
- Ripple: does not work
- Reactive: does not appear to work (edit: seems to work after changing primary color)
- Static: cannot change primary color
- Breathe: works and can change primary color
- Starlight: works, but cannot change primary color

Is there a better client to test with? Are there other non-keyboard functions I should test?

**Comment by flexa-rob** [comment]
@z3ntu - I have created a new fork and opened a new PR #2245

Yeah, I had to make changes to resolve merge conflicts from @dberlin 's fork.

Let me know if there are any changes I can make to resolve the above issues? (Ripple, Reactive, Static, change primary color)

**Comment by z3ntu** [comment]
Closing in favor of #2245, thanks!

**Pull request closed without merging**